src/net_processing.cpp

```diff
@@ -1542,6 +1542,14 @@ void static ProcessGetData(CNode* pfrom, const CChainParams& chainparams, CConnm
                 }
             }
             if (!push) {
+                if (!g_relay_txes && !pfrom->HasPermission(PF_RELAY)) {
+                    // If a blocks-only peer requests a tx and its not in our mapRelay, then disconnect them.
```

Comment:
But we'll have already tried relaying from the mempool and set push to true in that case, causing this code path to only work if we didn't have the tx at all? Am I missing something?

Reply:
no, you're not missing anything. You're completely right.

I think this logic needs to be moved up between the `if (mi != mapRelay.end())` block and `else if (pfrom->m_tx_relay->m_last_mempool_req.load().count())`. I've done that in https://github.com/bitcoin/bitcoin/compare/dc84c42a69557bff3d41baea498ceee822566423..dffa26f05fca414ecd2e72c6c9a7efc953d54651. Let me know what you think.